File: arduinobot_ws/src/arduinobot_cpp_examples/src/simple_parameter.cpp
// Incluye la librería principal de ROS 2 en C++
#include <rclcpp/rclcpp.hpp>

// Incluye el mensaje usado para el resultado de la callback de parámetros
#include <rcl_interfaces/msg/set_parameters_result.hpp>

// Librerías estándar de C++
#include <string>     // Para trabajar con cadenas std::string
#include <vector>     // Para manejar vectores de parámetros
#include <memory>     // Para punteros inteligentes (make_shared, SharedPtr)
#include <functional> // Para std::bind y std::placeholders

// Se usa std::placeholders::_1 para enlazar callbacks con un parámetro
using std::placeholders::_1;

// Definición de la clase SimpleParameter que hereda de rclcpp::Node
class SimpleParameter : public rclcpp::Node {
public:
  // Constructor de la clase
  SimpleParameter() : Node("simple_parameter") {
    // Declarar un parámetro entero con valor por defecto 28
    declare_parameter<int>("simple_int_param", 28);

    // Declarar un parámetro tipo string con valor por defecto "Antonio"
    declare_parameter<std::string>("simple_string_param", "Antonio");

    // Registrar el callback que se ejecutará cuando cambien parámetros
    // std::bind enlaza el método "paramChangeCallback" con el objeto actual (this)
    // "_1" representa el argumento que recibirá el callback (vector de parámetros cambiados)
    param_callback_handle_ = this->add_on_set_parameters_callback(
      std::bind(&SimpleParameter::paramChangeCallback, this, _1)
    );
  }

private:
  // Handle para mantener viva la callback de parámetros mientras el nodo exista
  rclcpp::Node::OnSetParametersCallbackHandle::SharedPtr param_callback_handle_;

  // Función que maneja los cambios de parámetros
  rcl_interfaces::msg::SetParametersResult
  paramChangeCallback(const std::vector<rclcpp::Parameter> &parameters) {
    // Crear el objeto de resultado que se devolverá
    rcl_interfaces::msg::SetParametersResult result;

    // Se asume que el cambio fue exitoso por defecto
    result.successful = true;

    // Iterar sobre todos los parámetros que cambiaron
    for (const auto &param : parameters) {
      // Obtener nombre y tipo del parámetro
      const auto &name = param.get_name();
      const auto type  = param.get_type();

      // Si el parámetro es "simple_int_param" y es de tipo entero
      if (name == "simple_int_param" &&
          type == rclcpp::ParameterType::PARAMETER_INTEGER) {
        // Imprimir en consola el nuevo valor
        RCLCPP_INFO(get_logger(),
                    "Param simple_int_param cambiado! Nuevo valor: %ld",
                    static_cast<long>(param.as_int()));

      // Si el parámetro es "simple_string_param" y es de tipo string
      } else if (name == "simple_string_param" &&
                 type == rclcpp::ParameterType::PARAMETER_STRING) {
        // Imprimir en consola el nuevo valor
        RCLCPP_INFO(get_logger(),
                    "Param simple_string_param cambiado! Nuevo valor: %s",
                    param.as_string().c_str());

      } else {
        // Si se intenta modificar un parámetro desconocido o con tipo inválido,
        // aquí podríamos rechazarlo poniendo: result.successful = false;
      }
    }

    // Retornar el resultado (éxito o fracaso)
    return result;
  }
};

// Función principal (punto de entrada del nodo)
int main(int argc, char *argv[]) {
  // Inicializa ROS 2 con los argumentos de línea de comandos
  rclcpp::init(argc, argv);

  // Crea una instancia del nodo SimpleParameter
  auto node = std::make_shared<SimpleParameter>();

  // Mantiene el nodo en ejecución escuchando callbacks y eventos
  rclcpp::spin(node);

  // Finaliza y libera los recursos de ROS 2
  rclcpp::shutdown();

  // Retorna 0 indicando ejecución exitosa
  return 0;
}

File: arduinobot_ws/src/arduinobot_py_examples/arduinobot_py_examples/simple_parameter.py
# Importa la librería principal de ROS2 para Python
import rclpy
# Importa la clase Node, que es la base para crear nodos en ROS2
from rclpy.node import Node
# Importa la estructura SetParametersResult, usada para indicar si un cambio de parámetro fue exitoso
from rcl_interfaces.msg import SetParametersResult
# Importa la clase Parameter, que representa parámetros dentro de un nodo
from rclpy.parameter import Parameter


# Definición de la clase SimpleParameter, que hereda de Node
class SimpleParameter(Node):
    def __init__(self):
        # Inicializa el nodo con el nombre "simple_paramaeter"
        super().__init__("simple_paramaeter")

        # Declara un parámetro entero llamado "simple_int_param" con valor por defecto 28
        self.declare_parameter("simple_int_param", 28)

        # Declara un parámetro tipo string llamado "simple_string_param" con valor por defecto "Anatonio"
        self.declare_parameter("simple_string_param", "Anatonio")

        # Registra un callback para manejar los cambios en los parámetros
        self.add_on_set_parameters_callback(self.paramChangeCallback)


    # Callback que se ejecuta cada vez que se intenta cambiar un parámetro
    def paramChangeCallback(self, params):
        # Objeto que indica si el cambio fue exitoso o no
        result = SetParametersResult()

        # Recorre todos los parámetros modificados
        for param in params:
            # Si el parámetro es "simple_int_param" y es de tipo entero
            if param.name == "simple_int_param" and param.type == Parameter.Type.INTEGER:
                # Imprime en consola el nuevo valor del parámetro
                self.get_logger().info("Param simple_int_param changed! New value is: %d" % param.value)
                # Marca el cambio como exitoso
                result.successful = True

            # Si el parámetro es "simple_string_param" y es de tipo string
            # (⚠️ Aquí hay un error: debería usarse param.type en lugar de param.type_)
            if param.name == "simple_string_param" and param.type == Parameter.Type.STRING:
                # Imprime en consola el nuevo valor del parámetro
                # (⚠️ Aquí también hay un error: %d es para enteros, debería ser %s para strings)
                self.get_logger().info("Param simple_string_param changed! New value is: %s" % param.value)

        # Devuelve el resultado del intento de cambio
        return result


# Función principal que arranca el nodo
def main():
    # Inicializa el sistema de ROS2
    rclpy.init()

    # Crea una instancia del nodo SimpleParameter
    simple_parameter = SimpleParameter()

    # Mantiene el nodo ejecutándose para escuchar cambios en parámetros
    rclpy.spin(simple_parameter)

    # Destruye el nodo al finalizar
    simple_parameter.destroy_node()

    # Apaga el sistema de ROS2
    rclpy.shutdown()


# Punto de entrada del script
if __name__ == '__main__':
    main()
